sklego/meta/_shrinkage_utils.py

```diff
@@ -0,0 +1,210 @@
+    """Weigh each group according to its size.
+
+    Parameters
+    ----------
+    group_sizes : array-like
+        The number of observations in each group.
+
+    Returns
+    -------
+    np.ndarray
+        The weights for each group.
+    """
+    return np.asarray(group_sizes)
+
+
+def min_n_obs_shrinkage(group_sizes, min_n_obs: int) -> np.ndarray:
+    """Use only the smallest group with a certain amount of observations.
+
+    Parameters
+    ----------
+    group_sizes : array-like
+        The number of observations in each group.
+    min_n_obs : int
+        The minimum number of observations for a group to be considered.
+
+    Returns
+    -------
+    np.ndarray
+        The weights for each group.
+    """
+    if min_n_obs > max(group_sizes):
+        raise ValueError(f"There is no group with size greater than or equal to {min_n_obs}")
+
+    res = np.zeros(len(group_sizes))
+    res[np.argmin(np.array(group_sizes) >= min_n_obs) - 1] = 1
+    return res
+
+
+def equal_shrinkage(group_sizes) -> np.ndarray:
+    """Each group is weighed equally.
+
+    Parameters
+    ----------
+    group_sizes : array-like
+        The number of observations in each group, must implement the `__len__` method.
+
+    Returns
+    -------
+    np.ndarray
+        The weights for each group.
+    """
+    return np.ones(len(group_sizes))
+
+
+class ShrinkageMixin:
+    """Mixin class for shrinkage functionality (setting shrinkage, checking shrinkage function, and fitting shrinkage
+    factors). The shrinkage factors are used to weigh the predictions of the different levels of a model.
+
+    Class inherits from this mixin should have the following attributes:
+
+    - `_ALLOWED_SHRINKAGE` : dict[str, callable]
+        A dictionary mapping the name of the shrinkage function to the function itself.
+    - `shrinkage` : str | callable | None
+        The shrinkage function to use. If a callable is passed, it should take an array of group sizes and return an
+        array of shrinkage factors.
+        `shrinkage` is parsed by `_set_shrinkage_function`, which then returns `shrinkage_function_` to be used in
+        `_fit_shrinkage_factors`.
+    - `shrinkage_kwargs` : dict[str, Any]
+        Additional keyword arguments to pass to the shrinkage function.
+    - `fitted_levels_` : list[str | int]
+        List of the levels that have been fitted.
+    - `estimators_` : dict[tuple[Any, ...], scikit-learn compatible estimator]
+    """
+
+    def _set_shrinkage_function(self):
+        """Set the shrinkage function and validate it if it is a custom callable"""
+        if isinstance(self.shrinkage, str) and self.shrinkage in self._ALLOWED_SHRINKAGE.keys():
+            shrinkage_function_ = self._ALLOWED_SHRINKAGE[self.shrinkage]
+
+        elif callable(self.shrinkage):
+            self.__check_shrinkage_func()
+            shrinkage_function_ = self.shrinkage
+
+        elif self.shrinkage is None:
```

Comment:
We discussed about adding proper comments in here. Please let me know if it is reasonable and clear enough